src/H5Dmpio.c

```diff
@@ -2583,20 +2633,25 @@ H5D__cmp_filtered_collective_io_info_entry(const void *filtered_collective_io_in
     addr2 = entry2->chunk_new.offset;
 
     /*
-     * If both chunk addresses are defined, H5_addr_cmp is safe to use.
-     * Otherwise, if both addresses aren't defined, compared chunk
-     * entries based on their chunk index. Finally, if only one chunk
-     * address is defined, return the appropriate value based on which
-     * is defined.
+     * If both chunk's file addresses are defined, H5_addr_cmp is safe to use.
+     * If only one chunk's file address is defined, return the appropriate
+     * value based on which is defined. If neither chunk's file address is
+     * defined, compare chunk entries based on their dataset object header
+     * address, then by their chunk index value.
      */
     if (H5_addr_defined(addr1) && H5_addr_defined(addr2)) {
         ret_value = H5_addr_cmp(addr1, addr2);
     }
     else if (!H5_addr_defined(addr1) && !H5_addr_defined(addr2)) {
-        hsize_t chunk_idx1 = entry1->index_info.chunk_idx;
-        hsize_t chunk_idx2 = entry2->index_info.chunk_idx;
+        haddr_t oloc_addr1 = entry1->index_info.dset_oloc_addr;
+        haddr_t oloc_addr2 = entry2->index_info.dset_oloc_addr;
+
+        if (0 == (ret_value = H5_addr_cmp(oloc_addr1, oloc_addr2))) {
+            hsize_t chunk_idx1 = entry1->index_info.chunk_idx;
+            hsize_t chunk_idx2 = entry2->index_info.chunk_idx;
 
-        ret_value = (chunk_idx1 > chunk_idx2) - (chunk_idx1 < chunk_idx2);
+            ret_value = (chunk_idx1 > chunk_idx2) - (chunk_idx1 < chunk_idx2);
```

Comment:
Couldn't two chunks from different datasets have the same index?

Reply:
Yes, but the logic above this should sort by dataset object header address first so that two chunks with the same index from different datasets will compare as different

Reply:
Oh wow not sure how I missed that